Content.Shared/GameObjects/Components/Doors/SharedDoorComponent.cs

```diff
@@ -1,26 +1,233 @@
-﻿using System;
+#nullable enable
+using System;
+using Robust.Shared.GameObjects;
+using Robust.Shared.GameObjects.Components;
+using Robust.Shared.IoC;
+using Robust.Shared.Interfaces.Timing;
 using Robust.Shared.Serialization;
+using Robust.Shared.ViewVariables;
+using Robust.Shared.Physics;
+using Robust.Shared.GameObjects.Components.Appearance;
 
 namespace Content.Shared.GameObjects.Components.Doors
 {
-    [NetSerializable]
-    [Serializable]
-    public enum DoorVisuals
+    public abstract class SharedDoorComponent : Component, ICollideSpecial
     {
-        VisualState,
-        Powered,
-        BoltLights
+        public override string Name => "Door";
+        public override uint? NetID => ContentNetIDs.DOOR;
+
+        [ViewVariables]
+        private DoorState _state = DoorState.Closed;
+        /// <summary>
+        /// The current state of the door -- whether it is open, closed, opening, or closing.
+        /// </summary>
+        public virtual DoorState State
+        {
+            get => _state;
+            protected set
+            {
+                if (_state == value)
+                {
+                    return;
+                }
+
+                _state = value;
+
+                Owner.EntityManager.EventBus.RaiseEvent(EventSource.Local, new DoorStateMessage(this, State));
+                SetAppearance(State switch
+                {
+                    DoorState.Open => DoorVisualState.Open,
+                    DoorState.Closed => DoorVisualState.Closed,
+                    DoorState.Opening => DoorVisualState.Opening,
+                    DoorState.Closing => DoorVisualState.Closing,
+                    _ => throw new ArgumentOutOfRangeException(),
+                });
```

Comment:
With proper door prediction this won't be needed and can save bandwidth as the client can infer it.

Reply:
I tried that, but it meant that the AppearanceComponent on the server wasn't making the change, and the `data` dictionary from the server's AppearanceComponent ended up overwriting the one on the client. I'm not sure how to fix this, although I didn't look into it particularly hard.

Reply:
If the server's appearance isn't being changed it shouldn't be being sent?